fix(cdk/visit/lambda_code): fix issue where register lambda does not

DillonRanwala opened the pull request:
update user registrations from website.

Problem:
Certain users have been unable to submit their user registrations from the emailed link. When they click submit, the page does not refresh because the register user lambda fails. Therefore, their user data is not put into the databases. The problem stems from the lambda trying to set the "last_updated" attribute to an empty string ("") for non-test users. However, this is a Number attribute in DynamoDB and cannot be set to an empty string.

Solution:
Changed the RegisterUser Lambda to avoid setting the "last_updated" attribute to an empty string. If the request json has that attribute (i.e. if it is a test request), it will set the attribute. Otherwise, the lambda will put the user info into DynamoDB without setting the "last_updated" attribute. This will allow both test and non-test registrations to pass without lambda errors.

Testing:
The Dev lambda updated with this change did not raise any errors with test or non-test json requests. Also, the dev website had successful registrations that put items into the DynamoDB table without issues.

Issue:
N/A, No open issue.

DillonRanwala committed:
fix(cdk/visit/lambda_code): fix issue where register lambda does not
update user registrations from website.

Problem:
Certain users have been unable to submit their user registrations from
the emailed link. When they click submit, the page does not refresh
because the register user lambda fails. Therefore, their user data is
not put into the databases. The problem stems from the lambda trying to
set the "last_updated" attribute to an empty string ("") for non-test
users. However, this is a Number attribute in DynamoDB and cannot be set
to an empty string.

Solution:
Changed the RegisterUser Lambda to avoid setting the "last_updated"
attribute to an empty string. If the request json has that attribute
(i.e. if it is a test request), it will set the attribute. Otherwise,
the lambda will put the user info into DynamoDB without setting the
"last_updated" attribute. This will allow both test and non-test
registrations to pass without lambda errors.

Testing:
The Dev lambda updated with this change did not raise any errors with
test or non-test json requests. Also, the dev website had successful
registrations that put items into the DynamoDB table without issues.

Issue:
N/A, No open issue.

ophilli commented on cdk/visit/lambda_code/register_user/register_user.py:
```diff
@@ -99,24 +99,43 @@ def add_user_info(self, user_info):
-
-            })
+        # if the json is from a test request it will have this ttl attribute
+        if "last_updated" in user_info:
+            user_table_response = self.dynamodbclient.put_item(
+                TableName=self.USERS_TABLE_NAME,
+                Item={
+                    'username': {'S': user_info['username']},
+                    'register_time': {'N': str(timestamp)},
+                    'first_name': {'S': user_info['firstName']},
+                    'last_name': {'S': user_info['lastName']},
+                    'gender': {'S': user_info['Gender']},
+                    'date_of_birth': {'S': user_info['DOB']},
+                    'position': {'S': user_info['UserPosition']},
+                    'grad_semester': {'S': grad_sem},
+                    'grad_year': {'S': grad_year},
+                    'majors': {'L': majors},
+                    'minors': {'L': minors},
+                    'last_updated':{"N":str(user_info['last_updated'])},   # get last_updated from test users, will set empty field for non-test users
+                })
+
+        # non-testing requests, will not set last_updated attribute
+        else:
+            user_table_response = self.dynamodbclient.put_item(
+                TableName=self.USERS_TABLE_NAME,
+                Item={
+                    'username': {'S': user_info['username']},
+                    'register_time': {'N': str(timestamp)},
+                    'first_name': {'S': user_info['firstName']},
+                    'last_name': {'S': user_info['lastName']},
+                    'gender': {'S': user_info['Gender']},
+                    'date_of_birth': {'S': user_info['DOB']},
+                    'position': {'S': user_info['UserPosition']},
+                    'grad_semester': {'S': grad_sem},
+                    'grad_year': {'S': grad_year},
+                    'majors': {'L': majors},
+                    'minors': {'L': minors},
+                })
```
Duplicating this `Item` dict just to add in the optional `last_updated` field seems liable to add in more places for bugs to hide. A discrepancy between the two dictionaries will be really hard to notice.

It would be better to create the dict, use the `if "last_updated" in user_info:` to decide whether or not to append `last_updated`, and then call `self.dynamodbclient.put_item()`.

DillonRanwala replied:
Added new commit with these changes.

ophilli reviewed (changes requested):
Nice catch! Left one comment with a requested change.

Based on your commit message it sounds like this was only noticed because users complained about their registrations failing. After you fix this PR & we get it merged in, it might be good to look at setting up an alarm on lambda failures https://www.google.com/search?q=cloudwatch+alarm+on+lambda+errors

DillonRanwala committed:
refactor(cdk/visit/lambda_code): create a single dict for the put_item()
operation for the user DynamoDB table in the Register User Lambda

Problem:
Needed to consolidate the put_item() dictionary into a single item
dictionary to avoid potential bugs with two copies of the dict.

Solution:
Created a single "user_table_item" dict that will be used by the
put_item() method.

Testing:

Dev lambda and tables did not show any errors.

Issue:
N/A, No open issue.
